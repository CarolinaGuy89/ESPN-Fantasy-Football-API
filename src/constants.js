/**
 * Maps `slotCategoryId`'s numerical enum to readable positions.
 * @type {Object}
 */
const slotCategoryIdToPositionMap = {
  0: 'QB',
  1: 'TQB',
  2: 'RB',
  3: 'RB/WR',
  4: 'WR',
  5: 'WR/TE',
  6: 'TE',
  7: 'OP',
  8: 'DT',
  9: 'DE',
  10: 'LB',
  11: 'DL',
  12: 'CB',
  13: 'S',
  14: 'DB',
  15: 'DP',
  16: 'D/ST',
  17: 'K',
  18: 'P',
  19: 'HC',
  20: 'Bench',
  21: 'IR',
  22: 'Unknown?', // TODO: Figure out what this is
  23: 'RB/WR/TE',
  24: 'Unknown?' // TODO: Figure out what this is
};

export { slotCategoryIdToPositionMap };
