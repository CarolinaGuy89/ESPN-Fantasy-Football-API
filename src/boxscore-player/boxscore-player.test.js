import _ from 'lodash';

import BaseObject from '../base-object/base-object.js';
import Player from '../player/player.js';

import BoxscorePlayer from './boxscore-player.js';

import { localObject, serverResponse } from './boxscore-player.stubs.js';

import { slotCategoryIdToPositionMap } from '../constants.js';

describe('BoxscorePlayer', () => {
  let boxscorePlayer;

  beforeEach(() => {
    boxscorePlayer = new BoxscorePlayer();
  });

  afterEach(() => {
    boxscorePlayer = null;
  });

  test('extends BaseObject', () => {
    expect(boxscorePlayer).toBeInstanceOf(BaseObject);
  });

  describe('attribute population from server response', () => {
    beforeEach(() => {
      boxscorePlayer = BoxscorePlayer.buildFromServer(serverResponse);
    });

    test('parses data correctly', () => {
      expect(boxscorePlayer).toMatchSnapshot();
    });
  });

  describe('attribute population from local object', () => {
    beforeEach(() => {
      boxscorePlayer = new BoxscorePlayer(localObject);
    });

    test('parses data correctly', () => {
      expect(boxscorePlayer).toMatchSnapshot();
    });
  });

  describe('responseMap', () => {
    describe('player', () => {
      describe('manualParse', () => {
        describe('when the passed data is undefined', () => {
          test('returns undefined', () => {
            const returnedPlayer = BoxscorePlayer.responseMap.player.manualParse();
            expect(returnedPlayer).toBeUndefined();
          });
        });

        describe('when the passed data is empty', () => {
          test('returns undefined', () => {
            const returnedPlayer = BoxscorePlayer.responseMap.player.manualParse({});
            expect(returnedPlayer).toBeUndefined();
          });
        });

        describe('when there is a cached player', () => {
          test('returns the cached player', () => {
            const playerId = 10;
            const cachedPlayer = Player.buildFromServer({ playerId });

            const returnedPlayer = BoxscorePlayer.responseMap.player.manualParse({ playerId });
            expect(returnedPlayer).toBe(cachedPlayer);

            Player.clearCache();
          });
        });

        describe('when there is not a cached player', () => {
          test('creates a new player', () => {
            const playerId = 10;
            Player.clearCache();

            const returnedPlayer = BoxscorePlayer.responseMap.player.manualParse({ playerId });
            expect(returnedPlayer).toBe(Player.get(playerId));

            Player.clearCache();
          });
        });
      });
    });

    describe('position', () => {
      describe('manualParse', () => {
        test('returns the mapped position', () => {
          const position = 2;

          const returnedPosition = BoxscorePlayer.responseMap.position.manualParse(position);
          expect(returnedPosition).toBe(_.get(slotCategoryIdToPositionMap, position));
        });
      });
    });

    describe('isLocked', () => {
      describe('manualParse', () => {
        describe('when valid enum key is passed', () => {
          test('switches on numerical enum correctly', () => {
            const lockedStatuses = {
              0: false,
              4: true
            };

            expect.hasAssertions();
            _.forEach(lockedStatuses, (value, key) => {
              const numKey = _.toNumber(key);
              const isLocked = BoxscorePlayer.responseMap.isLocked.manualParse(numKey);
              expect(isLocked).toBe(value);
            });
          });
        });

        describe('when invalid enum key is passed', () => {
          test('returns error string', () => {
            const isLocked = BoxscorePlayer.responseMap.isLocked.manualParse(2);
            expect(isLocked).toBeUndefined();
          });
        });
      });
    });
  });
});
