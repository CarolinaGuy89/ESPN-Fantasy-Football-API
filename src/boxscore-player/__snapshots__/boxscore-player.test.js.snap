// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxscorePlayer when creating a team from a server response parses and assigns data correctly 1`] = `
BoxscorePlayer {
  "isKeeper": false,
  "isLocked": true,
  "player": Player {
    "eligiblePositions": Array [
      "QB",
      "Bench",
    ],
    "firstName": "Aaron",
    "isActive": true,
    "isIREligible": false,
    "jerseyNumber": "12",
    "lastName": "Rodgers",
    "leagueId": undefined,
    "percentOwned": 99.21669,
    "percentOwnedChange": 0.23744,
    "percentStarted": 74.42904,
    "playerId": 8439,
    "seasonId": undefined,
  },
  "position": "QB",
  "projectedPoints": 20.46,
  "projectedStats": BoxscorePlayerPointStats {
    "lostFumbles": -0.47,
    "passing2PtConversions": 0.19,
    "passingInterceptions": -0.76,
    "passingTouchdowns": 7.87,
    "passingYards": 11.21,
    "rushing2PtConversions": 0.01,
    "rushingTouchdowns": 0.67,
    "rushingYards": 1.73,
  },
  "stats": BoxscorePlayerPointStats {
    "passingYards": 1.04,
  },
  "totalPoints": 1.04,
}
`;

exports[`BoxscorePlayer when creating a team locally parses and assigns data correctly 1`] = `
BoxscorePlayer {
  "isLocked": true,
  "player": Player {
    "eligiblePositions": Array [
      "QB",
      "Bench",
    ],
    "firstName": "Aaron",
    "isActive": true,
    "isIREligible": false,
    "jerseyNumber": "12",
    "lastName": "Rodgers",
    "leagueId": undefined,
    "percentOwned": 99.21669,
    "percentOwnedChange": 0.23744,
    "percentStarted": 74.42904,
    "playerId": 8439,
    "seasonId": undefined,
  },
  "position": "QB",
  "projectedPoints": 20.46,
  "projectedStats": BoxscorePlayerPointStats {
    "lostFumbles": -0.47,
    "passing2PtConversions": 0.19,
    "passingInterceptions": -0.76,
    "passingTouchdowns": 7.87,
    "passingYards": 11.21,
    "rushing2PtConversions": 0.01,
    "rushingTouchdowns": 0.67,
    "rushingYards": 1.73,
  },
  "stats": BoxscorePlayerPointStats {
    "passingYards": 1.04,
  },
  "totalPoints": 1.04,
}
`;
